File: scripts/_fetch_static_assets.py
import firebase_admin
from firebase_admin import credentials, firestore
import os
import requests
import json

# Configuration
SERVICE_ACCOUNT_KEY_PATH = './google_svs_account.json'  # Path to your Firebase service account key
FIREBASE_PROJECT_ID = 'mountaincats-61543'  # Your Firebase Project ID
# FIREBASE_STORAGE_BUCKET = 'mountaincats-61543.appspot.com' # Not strictly needed if only using Firestore and public URLs
CATS_COLLECTION = 'cats'
LOCAL_THUMBNAILS_DIR = 'public/images/thumbnails'
STATIC_DATA_JSON_PATH = 'src/lib/cats-static-data.json'
# ---

def initialize_firebase():
    """Initializes the Firebase Admin SDK."""
    if not os.path.exists(SERVICE_ACCOUNT_KEY_PATH):
        print(f"ERROR: Service account key not found at {SERVICE_ACCOUNT_KEY_PATH}.")
        print("Please ensure the path is correct and the file exists.")
        return None

    cred = credentials.Certificate(SERVICE_ACCOUNT_KEY_PATH)
    try:
        app = firebase_admin.initialize_app(
            credential=cred,
            options={'projectId': FIREBASE_PROJECT_ID},
            name='fetchStaticAssetsScript' # Unique name for this app instance
        )
        print(f"Firebase App '{app.name}' initialized successfully for project '{FIREBASE_PROJECT_ID}'.")
    except ValueError: # App already initialized
        app = firebase_admin.get_app(name='fetchStaticAssetsScript')
        print(f"Using existing Firebase App '{app.name}'.")
    return firestore.client(app=app)

def fetch_cats_data(db):
    """Fetches all documents from the cats collection in Firestore."""
    if not db:
        return []
    print(f"Fetching cat data from Firestore collection: '{CATS_COLLECTION}'...")
    cats_ref = db.collection(CATS_COLLECTION)
    docs = cats_ref.stream()
    
    all_cats_data = []
    for doc in docs:
        cat_data = doc.to_dict()
        cat_data['id'] = doc.id  # Ensure the document ID is included
        all_cats_data.append(cat_data)
    
    if not all_cats_data:
        print("No cat data found in Firestore.")
    else:
        print(f"Fetched {len(all_cats_data)} cat(s) from Firestore.")
    return all_cats_data

def download_and_update_thumbnails(cats_data):
    """
    Downloads thumbnails for each cat and updates the thumbnailUrl to a local path.
    Returns a new list with updated cat data.
    """
    os.makedirs(LOCAL_THUMBNAILS_DIR, exist_ok=True)
    print(f"Ensured local thumbnail directory exists: '{LOCAL_THUMBNAILS_DIR}'")

    updated_cats_list = []
    for cat in cats_data:
        cat_id = cat.get('id')
        original_thumbnail_url = cat.get('thumbnailUrl')
        cat_name = cat.get('name', f"ID: {cat_id}") # For logging

        updated_cat_entry = cat.copy() # Start with a copy

        if not cat_id:
            print(f"WARNING: Cat data missing 'id'. Skipping image processing for: {cat_name}")
            updated_cat_entry['thumbnailUrl'] = '' # Or some placeholder for missing ID
            updated_cats_list.append(updated_cat_entry)
            continue
        
        if not original_thumbnail_url or not original_thumbnail_url.startswith('http'):
            print(f"WARNING: Cat '{cat_name}' has an invalid or missing 'thumbnailUrl': '{original_thumbnail_url}'. Skipping download.")
            updated_cat_entry['thumbnailUrl'] = '' # Or keep original if it's a non-HTTP placeholder
            updated_cats_list.append(updated_cat_entry)
            continue

        try:
            print(f"Processing cat: '{cat_name}'. Original URL: '{original_thumbnail_url}'")
            response = requests.get(original_thumbnail_url, stream=True, timeout=15)
            response.raise_for_status()  # Raises HTTPError for bad responses (4XX or 5XX)

            # Determine file extension, default to .jpg
            file_extension = os.path.splitext(original_thumbnail_url.split('?')[0])[1]
            if not file_extension or len(file_extension) > 5 or len(file_extension) < 3: # Basic check
                file_extension = '.jpg'
            
            local_image_filename = f"{cat_id}{file_extension}"
            local_image_full_path = os.path.join(LOCAL_THUMBNAILS_DIR, local_image_filename)
            
            with open(local_image_full_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
            
            web_accessible_path = f"/{LOCAL_THUMBNAILS_DIR}/{local_image_filename}"
            print(f"Successfully downloaded and saved to: '{local_image_full_path}'. Web path: '{web_accessible_path}'")
            updated_cat_entry['thumbnailUrl'] = web_accessible_path

        except requests.exceptions.RequestException as e:
            print(f"ERROR: Downloading image for cat '{cat_name}' from {original_thumbnail_url}: {e}")
            updated_cat_entry['thumbnailUrl'] = '' # Or a placeholder indicating download failure
        except Exception as e:
            print(f"ERROR: An unexpected error occurred processing cat '{cat_name}': {e}")
            updated_cat_entry['thumbnailUrl'] = '' # Or a placeholder

        updated_cats_list.append(updated_cat_entry)
    return updated_cats_list

def save_static_data_json(cats_data):
    """Saves the list of cat data to a JSON file."""
    print(f"Saving updated cat data to JSON: '{STATIC_DATA_JSON_PATH}'...")
    try:
        with open(STATIC_DATA_JSON_PATH, 'w', encoding='utf-8') as f:
            json.dump(cats_data, f, ensure_ascii=False, indent=2)
        print(f"Successfully created static data JSON: '{STATIC_DATA_JSON_PATH}'")
    except IOError as e:
        print(f"ERROR: Could not write static data JSON to '{STATIC_DATA_JSON_PATH}': {e}")

def main():
    print("--- Starting Static Asset Fetching Process ---")
    db = initialize_firebase()
    if not db:
        print("Firebase initialization failed. Exiting.")
        return

    raw_cats_data = fetch_cats_data(db)
    if not raw_cats_data:
        print("No data fetched. Exiting.")
        return

    updated_cats_data = download_and_update_thumbnails(raw_cats_data)
    save_static_data_json(updated_cats_data)
    print("--- Static Asset Fetching Process Completed ---")

if __name__ == '__main__':
    # Ensure required libraries are installed:
    # pip install firebase-admin requests
    main()

File: scripts/fetch-static-assets.js
const admin = require('firebase-admin');
const fsPromises = require('fs').promises; // Using promises version of fs
const fs = require('fs'); // Standard fs for createWriteStream
const path = require('path');
const axios = require('axios'); // For downloading images

// --- Configuration ---
const SERVICE_ACCOUNT_KEY_PATH = './google_svs_account.json'; // Path to your Firebase service account key from project root
const FIREBASE_PROJECT_ID = 'mountaincats-61543'; // Your Firebase Project ID
const CATS_COLLECTION = 'cats';
const LOCAL_THUMBNAILS_DIR_RELATIVE = 'public/images/thumbnails'; // Relative to project root
const STATIC_DATA_JSON_PATH_RELATIVE = 'src/lib/cats-static-data.json'; // Relative to project root

// Absolute paths resolved from project root
const PROJECT_ROOT = path.resolve(__dirname, '..'); // Assuming script is in 'scripts' directory
const SERVICE_ACCOUNT_FULL_PATH = path.join(PROJECT_ROOT, SERVICE_ACCOUNT_KEY_PATH);
const LOCAL_THUMBNAILS_DIR = path.join(PROJECT_ROOT, LOCAL_THUMBNAILS_DIR_RELATIVE);
const STATIC_DATA_JSON_PATH = path.join(PROJECT_ROOT, STATIC_DATA_JSON_PATH_RELATIVE);
// ---

async function initializeFirebase() {
    console.log("Initializing Firebase Admin SDK...");
    try {
        // Check if service account key exists before requiring it
        await fsPromises.access(SERVICE_ACCOUNT_FULL_PATH, fs.constants.F_OK);
        const serviceAccountKey = require(SERVICE_ACCOUNT_FULL_PATH);

        if (admin.apps.length === 0) {
            admin.initializeApp({
                credential: admin.credential.cert(serviceAccountKey),
                projectId: FIREBASE_PROJECT_ID,
            });
            console.log(`Firebase App initialized successfully for project '${FIREBASE_PROJECT_ID}'.`);
        } else {
            console.log("Using existing Firebase App.");
        }
        return admin.firestore();
    } catch (error) {
        if (error.code === 'ENOENT') {
            console.error(`ERROR: Service account key not found at ${SERVICE_ACCOUNT_FULL_PATH}.`);
            console.error("Please ensure the path is correct and the file exists.");
        } else {
            console.error("ERROR: Firebase Admin SDK initialization failed.", error.message);
        }
        return null;
    }
}

async function fetchCatsData(db) {
    if (!db) return [];
    console.log(`Fetching cat data from Firestore collection: '${CATS_COLLECTION}'...`);
    try {
        const catsSnapshot = await db.collection(CATS_COLLECTION).get();
        const allCatsData = [];
        catsSnapshot.forEach(doc => {
            allCatsData.push({ id: doc.id, ...doc.data() });
        });
        if (allCatsData.length === 0) {
            console.log("No cat data found in Firestore.");
        } else {
            console.log(`Fetched ${allCatsData.length} cat(s) from Firestore.`);
        }
        return allCatsData;
    } catch (error) {
        console.error("Error fetching cats data from Firestore:", error);
        return [];
    }
}

async function downloadImage(url, localPath) {
    try {
        const response = await axios({
            method: 'GET',
            url: url,
            responseType: 'stream',
            timeout: 15000 // 15 seconds timeout
        });

        const writer = response.data.pipe(fs.createWriteStream(localPath));

        return new Promise((resolve, reject) => {
            writer.on('finish', resolve);
            writer.on('error', (err) => {
                console.error(`Error writing image to ${localPath}:`, err);
                reject(err);
            });
        });
    } catch (error) {
        if (error.response) {
            console.error(`HTTP error ${error.response.status} downloading ${url}`);
        } else {
            console.error(`Error downloading image ${url}:`, error.message);
        }
        throw error; // Re-throw to be caught by the caller
    }
}


async function downloadAndUpdateThumbnails(catsData) {
    console.log(`Ensuring local thumbnail directory exists: '${LOCAL_THUMBNAILS_DIR}'`);
    await fsPromises.mkdir(LOCAL_THUMBNAILS_DIR, { recursive: true });

    const updatedCatsList = [];

    for (const cat of catsData) {
        const catId = cat.id;
        const originalThumbnailUrl = cat.thumbnailUrl;
        const catName = cat.name || `ID: ${catId}`;
        let updatedCatEntry = { ...cat };

        if (!catId) {
            console.warn(`WARNING: Cat data missing 'id'. Skipping image processing for: ${catName}`);
            updatedCatEntry.thumbnailUrl = '';
            updatedCatsList.push(updatedCatEntry);
            continue;
        }

        if (!originalThumbnailUrl || !originalThumbnailUrl.startsWith('http')) {
            console.warn(`WARNING: Cat '${catName}' has an invalid or missing 'thumbnailUrl': '${originalThumbnailUrl}'. Skipping download.`);
            updatedCatEntry.thumbnailUrl = ''; 
            updatedCatsList.push(updatedCatEntry);
            continue;
        }

        try {
            console.log(`Processing cat: '${catName}'. Original URL: '${originalThumbnailUrl}'`);
            
            let fileExtension = '';
            try {
                fileExtension = path.extname(new URL(originalThumbnailUrl).pathname);
            } catch (urlError) {
                console.warn(`Could not parse URL to get extension for ${originalThumbnailUrl}. Defaulting extension.`);
            }
            
            if (!fileExtension || fileExtension.length > 5 || fileExtension.length < 3) { 
                fileExtension = '.jpg'; // Default extension
            }
            
            const localImageFilename = `${catId}${fileExtension}`;
            const localImageFullPath = path.join(LOCAL_THUMBNAILS_DIR, localImageFilename);
            
            await downloadImage(originalThumbnailUrl, localImageFullPath);
            
            // Refactored path generation
            const relativePath = path.relative(path.join(PROJECT_ROOT, 'public'), localImageFullPath);
            const standardizedPath = relativePath.replace(/\\/g, '/'); // Ensure forward slashes for web
            const webAccessiblePath = `/${standardizedPath}`;
            
            console.log(`Successfully downloaded and saved to: '${localImageFullPath}'. Web path: '${webAccessiblePath}'`);
            updatedCatEntry.thumbnailUrl = webAccessiblePath;

        } catch (error) {
            console.error(`Failed to download image for cat '${catName}'.`);
            updatedCatEntry.thumbnailUrl = ''; 
        }
        updatedCatsList.push(updatedCatEntry);
    }
    return updatedCatsList;
}

async function saveStaticDataJson(catsData) {
    console.log(`Saving updated cat data to JSON: '${STATIC_DATA_JSON_PATH}'...`);
    try {
        const jsonData = JSON.stringify(catsData, null, 2);
        await fsPromises.writeFile(STATIC_DATA_JSON_PATH, jsonData, 'utf-8');
        console.log(`Successfully created static data JSON: '${STATIC_DATA_JSON_PATH}'`);
    } catch (error) {
        console.error(`ERROR: Could not write static data JSON to '${STATIC_DATA_JSON_PATH}':`, error);
    }
}

async function main() {
    console.log("--- Starting Static Asset Fetching Process (Node.js) ---");
    
    const db = await initializeFirebase();
    if (!db) {
        console.error("Firebase initialization failed. Exiting.");
        process.exit(1); // Exit if Firebase can't be initialized
    }

    const rawCatsData = await fetchCatsData(db);
    if (rawCatsData.length === 0 && !(await db.collection(CATS_COLLECTION).limit(1).get()).empty) {
        // If fetchCatsData returns empty but the collection is not actually empty, it implies an error during fetch
        console.log("No data returned from fetchCatsData, possibly due to an error during fetch. Exiting.");
        process.exit(1);
    } else if (rawCatsData.length === 0) {
        console.log("No cat data found in Firestore. Process will complete, but no images will be downloaded.");
    }

    const updatedCatsData = await downloadAndUpdateThumbnails(rawCatsData);
    await saveStaticDataJson(updatedCatsData);
    console.log("--- Static Asset Fetching Process (Node.js) Completed ---");
}

if (require.main === module) {
    main().catch(error => {
        console.error("Unhandled error in main execution:", error);
        process.exit(1);
    });
}
